### PR TITLE
Implement MAC settings profile CRUD API

### PR DESCRIPTION
**Pull request by halimi**
#### Summary

References: https://github.com/TheThingsNetwork/lorawan-stack/issues/7380

This PR implements the MAC settings profile CRUD API. This is the first step (1 out of 6) of the proposed implementation plan.

#### Changes

  - New API for MAC settings profile CRUD operation

##### Steps

Pre-requisite: Create an application

Tests:
  - Create/Get/Update/List/Delete a MAC settings profile using the HTTP API 

##### Results

- Create a MAC settings profile:

Request:
```
 POST http://localhost:1885/api/v3/ns/applications/myapp/mac_settings_profiles
> Authorization: Bearer  MFRWG.QRQCFEX537L4S3ETCM56BLPQKTT2OGOSHRHUCXI.AKJG32COZUA2VD74Z7KOQMKWJPENCDT4KIHPAWYSHHTMKYJL4OIQ
> content-type: application/json
> data
{
  "mac_settings_profile": {
    "ids": {
      "application_ids": {
        "application_id": "myapp"
      },
      "profile_id": "prof-01"
    },
    "mac_settings": {
      "resets_f_cnt": false,
      "factory_preset_frequencies": [
        1,
        2,
        3
      ]
    }
  }
}
```

Response:
```
{
  "mac_settings_profile": {
    "ids": {
      "application_ids": {
        "application_id": "myapp"
      },
      "profile_id": "prof-01"
    },
    "mac_settings": {
      "factory_preset_frequencies": [
        "1",
        "2",
        "3"
      ],
      "resets_f_cnt": false
    }
  }
}
```
  
- Get a MAC settings profile

Request:
```
GET http://localhost:1885/api/v3/ns/applications/myapp/mac_settings_profiles/prof-01?field_mask=ids
> Authorization: Bearer MFRWG.QRQCFEX537L4S3ETCM56BLPQKTT2OGOSHRHUCXI.AKJG32COZUA2VD74Z7KOQMKWJPENCDT4KIHPAWYSHHTMKYJL4OIQ
```

Response:
```
{
  "mac_settings_profile": {
    "ids": {
      "application_ids": {
        "application_id": "myapp"
      },
      "profile_id": "prof-01"
    }
  }
}
```

- Update a MAC settings profile

Request:
```
PUT http://localhost:1885/api/v3/ns/applications/myapp/mac_settings_profiles/prof-01
> Authorization: Bearer MFRWG.QRQCFEX537L4S3ETCM56BLPQKTT2OGOSHRHUCXI.AKJG32COZUA2VD74Z7KOQMKWJPENCDT4KIHPAWYSHHTMKYJL4OIQ
> content-type: application/json
> data
{
  "mac_settings_profile": {
    "ids": {
      "application_ids": {
        "application_id": "myapp"
      },
      "profile_id": "prof-01"
    },
    "mac_settings": {
      "resets_f_cnt": true,
      "factory_preset_frequencies": [
        1,
        2,
        3,
        4,
        5
      ]
    }
  }
}
```

Response:
```
{
  "mac_settings_profile": {
    "ids": {
      "application_ids": {
        "application_id": "myapp"
      },
      "profile_id": "prof-01"
    },
    "mac_settings": {
      "factory_preset_frequencies": [
        "1",
        "2",
        "3",
        "4",
        "5"
      ],
      "resets_f_cnt": true
    }
  }
}
```

- List MAC settings profiles

Request:
```
GET http://localhost:1885/api/v3/ns/applications/myapp/mac_settings_profiles?field_mask=ids
> Authorization: Bearer MFRWG.QRQCFEX537L4S3ETCM56BLPQKTT2OGOSHRHUCXI.AKJG32COZUA2VD74Z7KOQMKWJPENCDT4KIHPAWYSHHTMKYJL4OIQ
```

Response:
```
{
  "mac_settings_profiles": [
    {
      "ids": {
        "application_ids": {
          "application_id": "myapp"
        },
        "profile_id": "prof-01"
      }
    },
    {
      "ids": {
        "application_ids": {
          "application_id": "myapp"
        },
        "profile_id": "prof-02"
      }
    }
  ]
}
```

- Delete a MAC settings profile

Request:
```
DELETE http://localhost:1885/api/v3/ns/applications/myapp/mac_settings_profiles/prof-01
> Authorization: Bearer MFRWG.QRQCFEX537L4S3ETCM56BLPQKTT2OGOSHRHUCXI.AKJG32COZUA2VD74Z7KOQMKWJPENCDT4KIHPAWYSHHTMKYJL4OIQ
```

Response:
```
{}
```

##### Regressions

...

#### Notes for Reviewers

...

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [x] Testing: The steps/process to test this feature are clearly explained including testing for regressions.
- [x] Infrastructure: If infrastructural changes (e.g., new RPC, configuration) are needed, a separate issue is created in the infrastructural repositories.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: api/ttn/lorawan/v3/networkserver_mac_settings_profile.proto [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  uint32 limit = 4 [(validate.rules).uint32.lte = 1000];
+  // Page number for pagination. 0 is interpreted as 1.
+  uint32 page = 5;
+}
+
+message ListMACSettingsProfilesResponse {
+  // The MAC settings profiles.
+  repeated MACSettingsProfile mac_settings_profiles = 1;
+  // The total number of MAC settings profiles that match the request.
+  uint32 total_count = 2;
```
Comment (johanstokking) [inline api/ttn/lorawan/v3/networkserver_mac_settings_profile.proto]: Please check that this is the convention. I think we often / always return the total count in a `x-total-count` gRPC header.

File: pkg/networkserver/grpc_mac_settings_profile.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+	"context"
+
+	"go.thethings.network/lorawan-stack/v3/pkg/auth/rights"
+	"go.thethings.network/lorawan-stack/v3/pkg/errors"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+)
+
+var (
+	errProfileAlreadyExists = errors.DefineAlreadyExists("mac_settings_profile_already_exists", "MAC settings profile already exists") // nolint: lll
+	errProfileNotFound      = errors.DefineNotFound("mac_settings_profile_not_found", "MAC settings profile not found")
```
Comment (johanstokking) [inline pkg/networkserver/grpc_mac_settings_profile.go]: nit; `errMACSettingsProfile...`

**Review by johanstokking — APPROVED** [review]
Great stuff!

File: api/ttn/lorawan/v3/networkserver_mac_settings_profile.proto (hunk 1 above)
Comment (halimi) [reply api/ttn/lorawan/v3/networkserver_mac_settings_profile.proto]: I've checked it and indeed we are using the `x-total-count` in the header so I've removed the `total_count` from the message.

File: pkg/networkserver/grpc_mac_settings_profile.go (hunk 2 above)
Comment (halimi) [reply pkg/networkserver/grpc_mac_settings_profile.go]: fixed

**Review by johanstokking — APPROVED** [review]

**Comment by KrishnaIyer** [comment]
Should we mark this API as experimental until we complete the rest of the ToDo items this is tested with end devices?

**Comment by johanstokking** [comment]
> Should we mark this API as experimental until we complete the rest of the ToDo items this is tested with end devices?

Yeah good call, let's do that.

**Comment by halimi** [comment]
> Should we mark this API as experimental until we complete the rest of the ToDo items this is tested with end devices?

@KrishnaIyer Can you please help me out? What should I do to mark it as experimental? Is it just the CHANGELOG or more?

**Comment by KrishnaIyer** [comment]
1. Changelog: Add a comment (This feature is experimental and subject to change)
2. Service comment in the proto file: NsMACSettingsProfileRegistry `EXPERIMENTAL: This service is subject to change.`

**Pull request merged**